argparse.SUPPRESS-like functionality for advanced uses

idoby opened the pull request.

lebrice commented:
Interesting!
Can you give me an example use-case for this? And how is this different than just parsing the class and then using `dataclasses.asdict` on the result?

idoby commented:
The use case is somewhat esoteric, I admit, but this is a feature that is available in argparse (and therefore SP) add_argument, so now it can be used with add_arguments as well.

I'm combining configs from multiple sources, so I need to know exactly what was specified on the command line, since the command line is neither the top priority or lowest priority source I have.
This way I can enable specifying each part of the dataclass from a different source and assemble the DC at the end of the entire process.

It sounds complicated but it makes keeping track of my ML experiments super easy.

lebrice reviewed (approved):
Looks fine to me. Let's see if all the tests pass.

Thanks for the PR!

idoby commented:
(It also affects whether the config MUST be specified on the cmdline, help text etc)

idoby commented:
I want to test it a little bit more with nested DCs before merging this

idoby commented:
OK seems to work fine. @lebrice I wrote a small hack when dealing with the boolean test cases there, is there a better way to accomplish this?

lebrice commented on test/test_default_args.py:
```diff
@@ -81,3 +82,103 @@ class SomeClass:
+    parser = ArgumentParser()
+    parser.add_arguments(SomeClass, dest='some_class', default=argparse.SUPPRESS)
+    if some_type == bool:
+        def strbool(s):
```
Use str2bool, wherever it's defined (probably in utils.py)

lebrice commented on test/test_default_args.py:
```diff
@@ -81,3 +82,103 @@ class SomeClass:
+
+        parser.add_argument('--c', type=strbool, default=argparse.SUPPRESS)
+    else:
+        parser.add_argument('--c', type=some_type, default=argparse.SUPPRESS)
```
Can you mix suppressed and non-suppressed arguments? What happens in that case?

idoby replied:
I've added another test case to demonstrate this

lebrice reviewed (approved):
Seems fine so far, and since it doesn't make any test fail, I don't have any objections to adding this.

Thanks @idoby !

idoby commented:
@lebrice I'm pretty happy with this now, I'd appreciate your review.

lebrice reviewed (approved):
Looks great! One last thing I think might be worthwhile to test would be to add a suppressed dataclass using `add_arguments`, as well as another dataclass with a non-suppressed default using `add_arguments`.

I don't think it's totally necessary for this to be merged, but if you find the time, I'd also appreciate a test for conflicts between two such argument groups (i.e. doing `parser.add_arguments(SomeClass, "foo", default=argparse.SUPPRESS)`, followed by `parser.add_arguments(SomeClass, "bar")` .

I believe this should bring out some bugs. Then again, you don't _have_ to add this test, but if you're willing to do it, I'd appreciate it.

Thanks again :)

idoby commented:
> add a suppressed dataclass using `add_arguments`, as well as another dataclass with a non-suppressed default using `add_arguments`.

> a test for conflicts between two such argument groups (i.e. doing `parser.add_arguments(SomeClass, "foo", default=argparse.SUPPRESS)`, followed by `parser.add_arguments(SomeClass, "bar")` .

Rebased onto master and added these tests. No bugs seem to have emerged this time.

idoby commented:
We can have a zoom call and I'll show you :)
